### DISCUSSION TITLE
Typescript gets updated even with configured SemVer and allowed version

### DISCUSSION THREAD

**Discussion by mavarazo** [renovatebot/renovate#14741] [Request Help] [answered]
Hello

First to say, that I'm new to Renovate and still try to find the best suitable configuration for the projects.
Renovate is suggesting a Typescript update, which I don't want, since we still use Angular 10.2.4 (allowed typescript versions: >=3.9.4 <=4.0.8).
Thanks for helping me, to find out why is Renovate doing this, and what I'm doing wrong :)

In the package.json of the project typescript is defined as: 
```
"devDependencies": {
    ...
    "typescript": "~4.0.7",
    ...
},
```

From the logs:
```
{
  "depType": "devDependencies",
  "depName": "typescript",
  "currentValue": "~4.0.7",
  "datasource": "npm",
  "prettyDepType": "devDependency",
  "lockedVersion": "4.0.8",
  "depIndex": 79,
  "updates": [
    {
      "bucket": "non-major",
      "newVersion": "4.6.2",
      "newValue": "~4.6.0",
      "releaseTimestamp": "2022-02-28T23:51:05.546Z",
      "newMajor": 4,
      "newMinor": 6,
      "updateType": "minor",
      "isRange": true,
      "branchName": "renovate/default"
    }
  ],
  "warnings": [],
  "versioning": "npm",
  "sourceUrl": "https://github.com/Microsoft/TypeScript",
  "homepage": "https://www.typescriptlang.org/",
  "currentVersion": "4.0.8",
  "isSingleVersion": false,
  "fixedVersion": "4.0.8"
}
```

From the preset:
```
{
  "extends": [
    ":semanticPrefixFixDepsChoreOthers",
    ":ignoreModulesAndTests",
    ":autodetectPinVersions",
    ":prHourlyLimit2",
    ":prConcurrentLimit10",
    "group:monorepos",
    "group:recommended",
    "workarounds:all",
    ":preserveSemverRanges"
  ],
  "packageRules": [
    {
      "matchPackagePatterns": ["^@angular/"],
      "matchUpdateTypes": ["major"],
      "enabled": false
    },
    {
      "matchPackageNames": ["^typescript"],
      "allowedVersions": "<=4.0.8"
    }
  ]
}
```

**Comment by viceice** [comment, marked as answer]
remove the `^` from `matchPackageNames`, it's a full string compare match field. PLease read the docs.

**Reply by viceice** [reply]
Instead of pinning typescript to a specific version, you should disable major and minor updates, so you always get patch updates.

**Reply by mavarazo** [reply]
Thanks, that fixed the issue.

Could you also explain why typescript get's an update even with `:preserveSemverRanges` and package.json defined
```
"devDependencies": {
    ...
    "typescript": "~4.0.7",
    ...
}
```

**Reply by rarkins** [reply]
`:preserveSemverRanges` means "don't pin". It doesn't mean "never ever update this".

**Reply by mavarazo** [reply]
Ah, I misunderstood `:preserveSemverRanges`. Thanks.
Probably last question :D: why do I have to make a exclusion for typescript, as we defined it with the tilde in `package.json`

**Reply by rarkins** [reply]
Renovate's job is to update dependencies. If you want that not to happen, you need an exclusion
